src/timely-util/src/operator.rs

```diff
@@ -462,13 +470,37 @@ where
         })
     }
 
-    fn expire_stream_at(&self, expiration: G::Timestamp) -> StreamCore<G, C1> {
-        self.inspect_container(move |data_or_frontier| {
-            if let Err(frontier) = data_or_frontier {
-                assert!(
-                    frontier.is_empty() || AntichainRef::new(frontier).less_than(&expiration),
-                    "frontier {frontier:?} has exceeded expiration {expiration:?}!",
-                );
+    fn expire_stream_at(
+        &self,
+        name: &str,
+        expiration: G::Timestamp,
+        token: Weak<()>,
+    ) -> StreamCore<G, C1> {
+        let name = format!("expire_stream_at({name})");
+        self.unary_frontier(Pipeline, &name.clone(), move |cap, _| {
+            let mut cap = Some(cap.delayed(&expiration));
```

Comment:
What is this delayed capability helping with?

Reply:
Added a comment.